### PR TITLE
MAM shaper touches

### PR DESCRIPTION
**Pull request by NelsonVides**
Storing names instead of recalculating list_to_atom again and again, also moving from `dict` to `maps`.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/121122) / small_tests / 2ceac51f45482df802269fe83149e1cbff5ac725
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3641/121122/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3641/121122/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-04-27_10.20.37/lib.mongooseim.logs/run.2022-04-27_10.20.38/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/121161) / small_tests / cc47bd239c1f9e96ee50c7698a493b1b7019f16e
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3641/121161/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3641/121161/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-04-27_10.24.08/lib.mongooseim.logs/run.2022-04-27_10.24.08/suite.log.html)

File: src/shaper_srv.erl [hunk 1]
```diff
@@ -57,46 +73,35 @@ child_spec(ProcName) ->
...
 wait(HostType, Domain, Action, FromJID, Size) ->
-    gen_server:call(select_worker(HostType, FromJID),
+    gen_server:call(select_worker(FromJID),
                     {wait, HostType, Domain, Action, FromJID, Size}).
 
 %% @doc Ask all shaper servers to forget current shapers and read settings again
-reset_all_shapers(HostType) ->
-    [reset_shapers(ProcName) || ProcName <- worker_names(HostType)].
+reset_all_shapers(_HostType) ->
+    [reset_shapers(ProcName) || {_N, ProcName} <- ets:tab2list(ejabberd_shaper_sup)].
```
Comment (arcusfelis) [inline src/shaper_srv.erl:104]: ideally, operations for storing and getting the list of workers should be in the same file.
i.e. ejabberd_shaper_sup should have a function get_worker_names.

File: src/shaper_srv.erl [hunk 2]
```diff
@@ -109,15 +114,15 @@ reset_shapers(ProcName) ->
 init(Args) ->
     State = #state{max_delay = proplists:get_value(max_delay, Args, 3000),
                    ttl = proplists:get_value(ttl, Args, 120),
-                   shapers = dict:new(),
-                   a_times = dict:new()
+                   shapers = #{},
+                   a_times = #{}
```
Comment (arcusfelis) [inline src/shaper_srv.erl:118]: oh, we don't need to store a_times separately.
Shaper has last_update field in the state. Ok, technically it is `last_update + delay`, but should work as well.

We also don't need to store shaper **at all**, if their value is `none` (but we can try to create them each time we want to update them). Even more, we probably could just do `mongoose_config:lookup_opt([shaper, Name])` before we call the shaper_srv and exit fast.


Also, the whole shaper logic could probably be moved to ETS, but that is another topic :D

**Review by arcusfelis — COMMENTED** [review]
use persistent terms.
+ some general info on shapers :)

File: src/shaper_srv.erl [hunk 3]
```diff
@@ -57,46 +73,35 @@ child_spec(ProcName) ->
...
+-spec start_link(atom()) -> ignore | {error, _} | {ok, pid()}.
 start_link(ProcName) ->
     gen_server:start_link({local, ProcName}, ?MODULE, [], []).
 
--spec worker_prefix() -> string().
-worker_prefix() ->
-    "ejabberd_shaper_".
+-spec worker_name(non_neg_integer()) -> atom().
+worker_name(N) ->
+    ets:lookup_element(ejabberd_shaper_sup, N, 2).
```
[the remark this answers is not in the document]
Comment (NelsonVides) [reply src/shaper_srv.erl:82]: Ok, I'll make it insanely fast then 😛

File: src/shaper_srv.erl (hunk 2 above)
Comment (NelsonVides) [reply src/shaper_srv.erl:118]: Ok, these comments are out of the scope of this quick PR, I wasn't trying to build better shapers, but just rework generic code to make it faster, if you can reword these details and make it all better, that'd be awesome 🙏🏽

**Review by arcusfelis — APPROVED** [review]

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/121208) / small_tests / 8433f2b7c7b27369683d5cb4bc99e492a9ab7bba
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3641/121208/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3641/121208/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-04-27_12.01.54/lib.mongooseim.logs/run.2022-04-27_12.01.55/suite.log.html)

**Pull request merged**